concordium-consensus/src/Concordium/GlobalState/Persistent/Account/StructureV1.hs

```diff
@@ -206,8 +211,42 @@ migratePersistentAccountStakeEnduring PersistentAccountStakeEnduringBaker{..} =
 migratePersistentAccountStakeEnduring PersistentAccountStakeEnduringDelegator{..} =
     return $! PersistentAccountStakeEnduringDelegator{..}
 
+-- | A monad transformer transformer that left-composes @StateT Amount@
+--  with a given monad transformer @t@. The purpose of this is to add state functionality for
+--  tracking the current active stake on an account, which is used when migrating an account
+--  between certain protocol versions.
+--
+--  A monad transformer transformer is used so that the 'lift' operation removes both the
+--  @StateT Amount@ and the underlying monad transformer @t@. This is important as the reference
+--  migration functions depend on using 'lift' to access the source block state.
+newtype StakedBalanceStateTT (t :: (Type -> Type) -> (Type -> Type)) (m :: Type -> Type) (a :: Type) = StakedBalanceStateTT
```

Comment:
what's the reason we need a transformer over a transformer, instead of just a normal monad transformer?

Reply:
Consider, for instance `migrateReference`:
```haskell
migrateReference ::
    forall t m ref1 ref2 a b.
    (MonadTrans t, Reference m ref1 a, Reference (t m) ref2 b) =>
    (a -> t m b) ->
    ref1 a ->
    t m (ref2 b)
migrateReference f hb = do
    a <- f =<< lift (refLoad hb)
    newRef <- refMake a
    (!newFlushedRef, _) <- refFlush newRef
    !_ <- lift (refUncache hb)
    return $! newFlushedRef
```

This relies on the monad `m` providing access to the source block state and `t m` providing access to the target block state. So `lift (refLoad hb)` loads from the old state and `refMake` creates a reference in the new state. When we add the extra `StateT` layer to our stack, we want `t = StateT . t0` and `m = m0` rather than `t = StateT` and `m = t0 m0`. So we need to transform the monad transformer.

Reply:
Makes sense, thanks!